In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

/its/home/nn268/.local/lib/python3.10/site-packages/pandas/core/arrays/masked.py:62: UserWarning: Pandas requires version '1.3.4' or newer of 'bottleneck' (version '1.3.2' currently installed).
  from pandas.core import (
/its/home/nn268/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.
  warnings.warn("Unable to import Axes3D. This may be due to multiple versions of "


repeating IDs pose a difficult situation.
file order stored local may not be the date order or other matching order.
Adding "B" to an ID does not help with merging the two DF

 full file name would be unique to each and can be matched upon.

 so going to include column for file name, and will use this and ID for the merge.

In [2]:
# folder paths

folder_jul18 = "July18/" # July18
folder_jul19 ="July19/"

folder_aug09 = "Aug09/"
folder_aug12 = "Aug12/"
folder_aug13 = "Aug13/"
folder_aug14 = "Aug14/"
folder_aug15 = "Aug15/"
folder_aug16 = "Aug16/"
folder_aug21 = "Aug21/"
folder_aug23 = "Aug23/"
folder_aug27 = "Aug27/"
folder_aug28 = "Aug28/"
folder_aug30 = "Aug30/"

folder_sep09 = "Sept09/"
folder_sep10 = "Sept10/"
fodler_sep11 = "Sept11/"
folder_sep20 = "Sept20/"
folder_sep23 ="Sept23/"
folder_sep24 = "Sept24/"
folder_sep26 ="Sept26/"

In [3]:
new_nest = pd.read_csv('NewNestTidy25.csv')#("AbbotsWood2024_newnest_tidy.csv")
wigwam_nest = pd.read_csv("AbbotsWood2024_wigwam_tidy.csv")
#nest_conditions = pd.read_csv("AbbotsWood2024_newnest_tidy2.csv") # has nest info, both wig and NN

In [4]:
print(list(new_nest))

new_nest = new_nest[['Date', 'ID', 'Condition', 'Sector', 'Direction']]

print(list(new_nest))


['Date', 'ID', 'Sector', 'Direction', 'Condition']
['Date', 'ID', 'Condition', 'Sector', 'Direction']


need to add nest type to the above DF

In [5]:
print(list(new_nest))
print("----")
print(np.unique(new_nest['Condition']))
newnest_str = "NewNest"
new_nest.insert(loc=2, column="Nest", value=newnest_str)
new_nest['Nest'].fillna(newnest_str, inplace=True)
new_nest = new_nest.replace('Unfamiliar ZV 3 ', 'Unfamiliar ZV 3')
print(np.unique(new_nest['Condition']))

print("----")
print(list(new_nest))
print("------------------------------------------------------")
print(list(wigwam_nest))
print("----")
print(np.unique(wigwam_nest['Condition']))
wigwam_str = "Wigwam"
wigwam_nest.insert(loc=2, column="Nest", value=wigwam_str)
wigwam_nest['Nest'].fillna(wigwam_str, inplace=True)
print("----")
print(list(wigwam_nest))

print(np.unique(wigwam_nest['Condition']))

['Date', 'ID', 'Condition', 'Sector', 'Direction']
----
['Minus 3m ZV' 'Off-route 1 ZV' 'On-route 1 ZV' 'On-route 1B ZV'
 'On-route 2 ZV' 'Plus 3m ZV' 'Route 0m ZV' 'Unfamiliar FV'
 'Unfamiliar ZV' 'Unfamiliar ZV 2' 'Unfamiliar ZV 3']
['Minus 3m ZV' 'Off-route 1 ZV' 'On-route 1 ZV' 'On-route 1B ZV'
 'On-route 2 ZV' 'Plus 3m ZV' 'Route 0m ZV' 'Unfamiliar FV'
 'Unfamiliar ZV' 'Unfamiliar ZV 2' 'Unfamiliar ZV 3']
----
['Date', 'ID', 'Nest', 'Condition', 'Sector', 'Direction']
------------------------------------------------------
['Date', 'ID', 'Sector', 'Direction', 'Condition']
----
['Off-route 1 ZV' 'On-route 1 ZV' 'On-route 2 ZV' 'Unfamiliar FV'
 'Unfamiliar ZV']
----
['Date', 'ID', 'Nest', 'Sector', 'Direction', 'Condition']
['Off-route 1 ZV' 'On-route 1 ZV' 'On-route 2 ZV' 'Unfamiliar FV'
 'Unfamiliar ZV']


In [6]:
bothNests = pd.concat([new_nest, wigwam_nest], ignore_index=True)

In [7]:
np.unique(bothNests['Nest'])

array(['NewNest', 'Wigwam'], dtype=object)

In [8]:
ids = np.unique(bothNests['ID'].astype('str'))
ids = ids.astype('str')
print(np.unique(bothNests['Condition']))

bothNests_unf2 = bothNests[bothNests['Condition'] == 'Unfamiliar ZV 2']
print(len(bothNests_unf2))

['Minus 3m ZV' 'Off-route 1 ZV' 'On-route 1 ZV' 'On-route 1B ZV'
 'On-route 2 ZV' 'Plus 3m ZV' 'Route 0m ZV' 'Unfamiliar FV'
 'Unfamiliar ZV' 'Unfamiliar ZV 2' 'Unfamiliar ZV 3']
61


In [9]:
bothNests[bothNests['ID'].astype('str')=='193']

,Date,ID,Nest,Condition,Sector,Direction


In [10]:
#print(ids)

bad_ids_2 = []
bad_ids_1 = []
for i in ids:
    #print(len(i))
    if len(i) <=1:
        bad_ids_1.append(i)
    elif len(i) <=2:
            bad_ids_2.append(i)


print(bad_ids_2)

['10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57', '58', '59', '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70']


In [11]:
bothNests['ID'] = bothNests['ID'].astype('str')

for idx, row in bothNests.iterrows():
    #print(row['ID'])
    if len(row['ID']) <=1:
        #print(row['ID'])
        bothNests['ID'][idx] = f"00{row['ID']}"
        
    elif len(row['ID']) <= 2:
        bothNests['ID'][idx] = f"0{row['ID']}"
    else:
        continue

for i in bothNests['ID']:
    if len(i) <=2:
        print(i)

In [12]:
for i in bothNests['ID']:
    #print(i)
    if len(i) <=1:
        i = f'00{i}'
    elif len(i) <=2:
            i = f'0{i}'


for i in bothNests['ID']:
    print(i)

561
562
563
564
565
566
579
580
581
582
583
584
590
591
592
593
594
611
612
613
614
615
635
636
637
638
639
640
641
663
664
665
666
685
686
687
688
394
395
694
695
696
697
698
721
722
723
724
725
726
728
729
730
731
732
772
773
774
775
776
777
778
779
780
781
552
553
554
556
557
558
573
574
575
576
577
578
585
586
587
588
589
605
606
607
608
609
610
621
627
628
629
630
631
632
634
659
660
661
662
690
691
692
693
699
700
701
702
703
715
716
717
718
719
720
727
733
734
735
736
737
738
739
740
741
742
743
744
745
746
747
748
749
567
568
569
570
571
572
595
596
597
598
599
600
601
602
603
604
616
617
618
619
620
622
623
624
625
626
653
654
655
656
657
658
680
681
682
683
684
704
705
706
707
708
750
751
752
753
754
755
756
757
758
759
760
761
762
763
764
765
766
767
768
769
770
771
642
643
644
645
646
647
648
649
650
651
652
667
668
669
670
671
672
673
674
675
676
677
678
679
709
710
711
712
713
782
783
784
785
786
787
788
789
792
793
794
795
796
797
798
799
800
801
802
803
804
805
806
807


In [13]:


def merging2onedf(folder, **kwargs):
    # for reading in csvs.
    # turning to pandas df
    # option to add additional cols
    
    used_id_list = []
    df_list = []
    #alphabet = ["B", "C", "D", "E", "F"] 
    for file in os.listdir(folder):
        if 'csv' in file:
            #if file[5:8] in used_id_list: # Check if the vid ID has already been used
            #    fileID = file[5:8]+"B"
            #    print(fileID) # show which IDs this happened to
            #else:
            if file[0] == 'A' or file[0]== 'B':
                additional = file[0]
            else:
                additional = ""
            fileID = file[5:8]+additional
            df = pd.read_csv(folder+file)
            df.insert(loc=0, column='ID', value=fileID)
            df['ID'].fillna(fileID, inplace=True)
            for key, val in kwargs.items():
                df.insert(loc=0, column=key, value=val)
                # fill all rows with the same val
                df[key].fillna(val, inplace=True)
            df_list.append(df)

            used_id_list.append(fileID)

                
    df_all = pd.concat(df_list)
    return df_all



In [14]:
df_jul18 = merging2onedf(folder_jul18, Date = "Jul-18")
df_jul19 = merging2onedf(folder_jul19, Date = "Jul-19")

df_aug09 = merging2onedf(folder_aug09, Date ="Aug-09")
df_aug12 = merging2onedf(folder_aug12, Date ="Aug-12")
df_aug13 = merging2onedf(folder_aug13, Date ="Aug-13")
df_aug14 = merging2onedf(folder_aug14, Date ="Aug-14")
df_aug15 = merging2onedf(folder_aug15, Date ="Aug-15")
df_aug16 = merging2onedf(folder_aug16, Date ="Aug-16")

df_aug21 = merging2onedf(folder_aug21, Date ="Aug-21")
df_aug23 = merging2onedf(folder_aug23, Date ="Aug-23")
df_aug27 = merging2onedf(folder_aug27, Date ="Aug-27")
df_aug28 = merging2onedf(folder_aug28, Date ="Aug-28")
df_aug30 = merging2onedf(folder_aug30, Date ="Aug-30")
df_sep09 = merging2onedf(folder_sep09, Date ="Sep-09")
df_sep10 = merging2onedf(folder_sep10, Date ="Sep-10")
df_sep11 = merging2onedf(fodler_sep11, Date ="Sep-11")
df_sep20 = merging2onedf(folder_sep20, Date ="Sep-20")
df_sep23 = merging2onedf(folder_sep23, Date ="Sep-23")
df_sep24 = merging2onedf(folder_sep24, Date ="Sep-24")
df_sep26 = merging2onedf(folder_sep26, Date ="Sep-26")

In [15]:
df_sep23_192 = df_sep23[df_sep23['ID'].astype('str') == '192']

print(df_sep23_192)

       Date   ID  frame           x          y
0    Sep-23  192      0   -7.874331  -5.678055
1    Sep-23  192      1   -7.179074  -5.738707
2    Sep-23  192      2   -6.876221  -5.762784
3    Sep-23  192      3   -6.972239  -5.692476
4    Sep-23  192      4   -6.971978  -5.744445
..      ...  ...    ...         ...        ...
666  Sep-23  192    670  282.664422  35.611265
667  Sep-23  192    671  283.220257  34.944740
668  Sep-23  192    677  285.552246  31.396012
669  Sep-23  192    678  285.628006  31.217196
670  Sep-23  192    679  286.015017  31.055530

[671 rows x 5 columns]


 Wigwam Repeats:
 376
 375
 273

 Repeat over both nests:

 395
 
 394

In [16]:
df_alltogether = pd.concat([df_jul18,df_jul19,df_aug09,df_aug12,df_aug13,df_aug14,df_aug15,df_aug16,df_aug21,df_aug23,df_aug27,df_aug28,df_aug30,
                            df_sep09,df_sep10,df_sep11,df_sep20,df_sep23,df_sep24,df_sep26], ignore_index=True)

In [17]:
list(df_alltogether)

['Date', 'ID', 'frame', 'x', 'y']

In [18]:
list(df_alltogether)

print(type(df_alltogether['ID'].iloc[0]))



<class 'str'>


In [19]:
print(len(np.unique(df_alltogether['ID'].astype('str'))))

907


In [20]:
all_rawData = [df_jul18,df_jul19,df_aug09,df_aug12,df_aug13,df_aug14,df_aug15,df_aug16,df_aug21,df_aug23,df_aug27,df_aug28,df_aug30,
                            df_sep09,df_sep10,df_sep11,df_sep20,df_sep23,df_sep24,df_sep26]


for i in all_rawData:
    print(len(np.unique(i['ID'].astype('str'))))

39
42
15
57
30
31
15
22
30
37
69
33
105
108
62
61
32
59
57
22


In [21]:
allIDS =  np.unique(df_alltogether['ID'].astype('str'))



Number of IDS for each day of data collection


df_jul18  :  39

df_jul19  : 42

df_aug09  : 15

df_aug12  : 57

df_aug13  : 30

df_aug14  : 31

df_aug15  : 15

df_aug16  :  22

df_aug21  :  30

df_aug23  :  37

df_aug27  :  69

df_aug28  : 33

df_aug30  :  105

df_sep09 :  108

df_sep10 :  62

df_sep11 :  61

df_sep20 :  31

df_sep23 :  59

df_sep24 :  57

df_sep26 :  22

In [22]:
df_alltogether.head
print(type(df_alltogether['ID'].iloc[0]))


#nest_conditions['ID'] = nest_conditions['ID'].astype('str')
#print(type(nest_conditions['ID'].iloc[0]))

<class 'str'>


In [23]:
np.unique(df_alltogether['ID'].astype('str'))

array(['001', '002', '003', '004', '005', '006', '007', '009', '011',
       '012', '013', '014', '015', '016', '017', '018', '019', '020',
       '021', '022', '023', '024', '025', '026', '027', '028', '029',
       '030', '031', '032', '033', '034', '035', '036', '037', '038',
       '039', '040', '041', '042', '043', '044', '045', '046', '047',
       '048', '049', '050', '051', '052', '053', '054', '055', '056',
       '057', '058', '059', '060', '061', '062', '063', '064', '065',
       '066', '067', '068', '069', '070', '108', '109', '110', '111',
       '112', '113', '114', '115', '116', '117', '118', '119', '120',
       '121', '122', '123', '124', '125', '126', '127', '128', '129',
       '130', '131', '132', '133', '134', '135', '136', '137', '138',
       '139', '142', '143', '146', '147', '148', '149', '150', '151',
       '152', '153', '155', '156', '157', '158', '159', '160', '161',
       '162', '163', '165', '166', '167', '168', '169', '170', '171',
       '172', '173',

In [24]:
bothNests

,Date,ID,Nest,Condition,Sector,Direction
0,Aug-21,561,NewNest,Off-route 1 ZV,1,0
1,Aug-21,562,NewNest,Off-route 1 ZV,6,112.5
2,Aug-21,563,NewNest,Off-route 1 ZV,3,45
3,Aug-21,564,NewNest,Off-route 1 ZV,9,180
4,Aug-21,565,NewNest,Off-route 1 ZV,16,337.5
...,...,...,...,...,...,...
870,Aug-16,531,Wigwam,Unfamiliar FV,4,67.5
871,Aug-16,532,Wigwam,Unfamiliar FV,2,22.5
872,Aug-16,534,Wigwam,Unfamiliar FV,2,22.5
873,Aug-16,536,Wigwam,Unfamiliar FV,6,112.5


In [25]:
print(bothNests[bothNests['ID'].astype('str') == '192'])

Empty DataFrame
Columns: [Date, ID, Nest, Condition, Sector, Direction]
Index: []


In [26]:
bothNests_unf2 = bothNests[bothNests['Condition'] == 'Unfamiliar ZV 2']
bothNests_unf3 = bothNests[bothNests['Condition'] == 'Unfamiliar ZV 3']
#df_alltogether_unf2 = df_alltogether[df_alltogether['Condition'] == 'Unfamiliar ZV 2']

print(len(np.unique(bothNests_unf2['ID'])))

print(len(np.unique(bothNests_unf3['ID'])))

#print(len(np.unique(df_alltogether_unf2['ID'])))

61
63


In [27]:
bothNests['ID'] = bothNests['ID'].astype('str')
df_alltogether['ID'] = df_alltogether['ID'].astype('str')

In [28]:
np.unique(bothNests['Nest'])

array(['NewNest', 'Wigwam'], dtype=object)

In [29]:
df_192 = df_alltogether[df_alltogether['ID']==192]

print(df_192)

Empty DataFrame
Columns: [Date, ID, frame, x, y]
Index: []


In [30]:
print(list(df_alltogether))
print()
print(list(bothNests))


IDS_df_alltogether = np.unique(df_alltogether['ID'])

IDS_bothNests = np.unique(bothNests['ID'])


print(len(IDS_df_alltogether))


print(len(IDS_bothNests))


missingIDS = []
for i in IDS_df_alltogether:
    if i not in IDS_bothNests:
        missingIDS.append(i)



print(missingIDS)

['Date', 'ID', 'frame', 'x', 'y']

['Date', 'ID', 'Nest', 'Condition', 'Sector', 'Direction']
907
873
['192', '193', '194', '195', '196', '197', '198', '199', '200', '201', '202', '203', '204', '205', '236', '237', '238', '239', '240', '241', '242', '243', '244', '245', '246', '247', '248', '249', '250', '251', '252', '253', '254', '255', '256', '257', '258', '259', '260', '261', '262', '273', '279', '284']


'192'

'193'

'194'

'195'
'196'
'197'
'198'
'199'
'200'
'201'
'202'
'203'
'204'
'205'
'236'
'237'
'238'
'239'
'240'
'241'
'242'
'243'
'244'
'245'
'246'
'247'
'248'
'249'
'250'
'251'
'252'
'253'
'254'
'255'
'256'
'257'
'258'
'259'
'260'
'261'
'262'
'273A'
'273B'
'279'
'284'
'375A'
'375B'
'376A'
'376B'
'694'
'695

In [31]:
bothNests['ID'] = bothNests['ID'].astype('str')
df_alltogether['ID'] = df_alltogether['ID'].astype('str')

In [32]:
np.unique(bothNests['Date'])

array(['Aug-09', 'Aug-12', 'Aug-13', 'Aug-14', 'Aug-15', 'Aug-16',
       'Aug-21', 'Aug-23', 'Aug-27', 'Aug-28', 'Aug-30', 'Jul-18',
       'Jul-19', 'Sep-09', 'Sep-10', 'Sep-11', 'Sep-20', 'Sep-23',
       'Sep-24'], dtype=object)

In [33]:
np.unique(df_alltogether['Date'])

array(['Aug-09', 'Aug-12', 'Aug-13', 'Aug-14', 'Aug-15', 'Aug-16',
       'Aug-21', 'Aug-23', 'Aug-27', 'Aug-28', 'Aug-30', 'Jul-18',
       'Jul-19', 'Sep-09', 'Sep-10', 'Sep-11', 'Sep-20', 'Sep-23',
       'Sep-24', 'Sep-26'], dtype=object)

In [34]:
df_alltogether2 = df_alltogether.merge(bothNests,how='inner', on=['Date','ID'])

In [35]:
list(df_alltogether2)

['Date', 'ID', 'frame', 'x', 'y', 'Nest', 'Condition', 'Sector', 'Direction']

In [36]:
np.unique(df_alltogether2['Nest'])

array(['NewNest', 'Wigwam'], dtype=object)

In [37]:
print(bothNests[bothNests['ID'].astype('str')=='192'])


Empty DataFrame
Columns: [Date, ID, Nest, Condition, Sector, Direction]
Index: []


In [38]:
print(df_alltogether2[df_alltogether2['ID']=='192'])

Empty DataFrame
Columns: [Date, ID, frame, x, y, Nest, Condition, Sector, Direction]
Index: []


In [39]:
df_alltogether2['Nest'] =df_alltogether2['Nest'].astype('str')

np.unique(df_alltogether2['Nest'])

array(['NewNest', 'Wigwam'], dtype=object)

In [40]:
df_alltogether2['Condition'] = df_alltogether2['Condition'].astype('str')
np.unique(df_alltogether2['Condition'])

array(['Minus 3m ZV', 'Off-route 1 ZV', 'On-route 1 ZV', 'On-route 1B ZV',
       'On-route 2 ZV', 'Plus 3m ZV', 'Route 0m ZV', 'Unfamiliar FV',
       'Unfamiliar ZV', 'Unfamiliar ZV 2', 'Unfamiliar ZV 3'],
      dtype=object)

xs = []
for idx, i in df_alltogether2.iterrows():
    xs.append(str(i))

print(np.unique(xs))

In [41]:
df_alltogether2.to_csv("WhereTheNansGone.csv")

In [42]:
np.unique(df_alltogether2['Nest'])

array(['NewNest', 'Wigwam'], dtype=object)

In [43]:
df_alltogether2.to_csv("AbbottsWoodsCombinedTidy.csv")